File: src/source_v0.2_Java6-8/SmallBoxForFansub/docs/advkaraoke.py
#!/usr/bin/env jython

from smallboxforfansub.scripting import ScriptPlugin
from smallboxforfansub.scripting import InputDialog

# Register the function in the main program.
ScriptPlugin.pythonRegister("myname", "myfunction", "1.0", "no description", "author", "", "")

# Register the function in the main program.
ScriptPlugin.pythonRegister("myname2", "myfunction2", "1.0", "no description", "author", "", "")

# Register the function in the main program.
ScriptPlugin.pythonRegister("Lightning (array mode)", "lightning", "1.0",
"Lightning effect with the font PowerOfGod.\nFrom the original Lua script for Aegisub of April 2009.", "AssFxMaker owner (Chien-Rouge)", "", "")

# Register the function in the main program.
ScriptPlugin.addTheme("5B5B5B", "999999", "DeepGray")

def myfunction():
    print "Hello World !! A (jython)"



def myfunction2():
    print "Hello World !! B (jython)"
    # Test
    ScriptPlugin.addResLine("Dialogue: 0,0:01:23.77,0:01:29.67,Default,,0000,0000,0000,,Je suis là !")


def lightning():
    # Get selected line (line = head + sentence)
    lines = ScriptPlugin.getSelectedOrgLines()
    
    # Ask for size of the video
    dialog = InputDialog(None,True)
    dialog.setLocationRelativeTo(None)
    dialog.setDialogMessage("Choose the video width to set up the karaoke.")
    dialog.setDialogTitle("Choose a size.")
    videowidth = dialog.showDialog("1280") # string
    dialog.setDialogMessage("Choose the video height to set up the karaoke.")
    dialog.setDialogTitle("Choose a size.")
    videoheight = dialog.showDialog("720") # string
    
    posx = str(int(videowidth)/2)
    posy = str(int(videoheight)/18)
    
    for line in lines:
        # Get the sentence for this line
        sentence = ScriptPlugin.getSentence(line)
        
        # Get the head for this line
        head = ScriptPlugin.getHead(line)

        # Get a table of syllable parameters for the basic karaoke
        # [i][0] : syllable                   string
        # [i][1] : hundredth                  integer
        # [i][2] : thousandth                 integer
        # [i][3] : thousandth (start)         integer
        # [i][4] : thousandth (end)           integer
        osyl = ScriptPlugin.phKaraoke(sentence)
        
        ScriptPlugin.regenerateAssStyleCollection()
        assstyle = ScriptPlugin.getAssStyle(line)
        linewidth = ScriptPlugin.phXSentenceWidth(osyl, assstyle) # integer
        pxArray = ScriptPlugin.phPosXSyllable(osyl, assstyle) # integer
        
        leftstartpos = (int(videowidth)-linewidth)/2
        
        # ======================================================================
        # Layer 1 : Disappearance of syllable when pronounced
        # ======================================================================
    
        # Create a new sentence
        newsentence = ""

        # Counter
        counter = 0

        # Loop onto each syllable to create the sentence
        while counter<len(osyl):

            syl = osyl[counter][0]
            sstart = str(osyl[counter][3])              # start become a string

            # The first syllable may contains overrides
            if counter==0:
                newsentence = "{\\an5\\pos("+posx+","+posy+")\\alpha&HFF&}"+syl
            # All others syllables
            else:
                newsentence = newsentence + "{\\r\\t("+sstart+","+sstart+",\\alpha&HFF&)}"+syl

            counter += 1

        # Reformat the line        
        line = ScriptPlugin.getAssLineOf(head, newsentence)

        # Add a new line to the table of the main program.
        ScriptPlugin.addResLine(line)
        
        # ======================================================================
        # Layer 2 : Appearance of syllable when pronounced
        # ======================================================================
    
        # Create a new sentence
        newsentence = ""

        # Counter
        counter = 0

        # Loop onto each syllable to create the sentence
        while counter<len(osyl):

            syl = osyl[counter][0]
            smiddle = str(osyl[counter][3]+osyl[counter][2]/2)       # middle become a string
            sstart = str(osyl[counter][3])                           # start become a string
            send = str(osyl[counter][4])                             # end become a string

            # The first syllable may contains overrides
            if counter==0:
                newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")\\fscx200\\fscy200\\1c&HFFD700&"
                newsentence = newsentence + "\\t("+send+","+str(osyl[counter][4]+600)+",0.5,\\fscx100\\fscy100\\1c&HFFFFFF&)}"+syl
                line = ScriptPlugin.getAssLineOf(head, newsentence) # Reformat the line
                ScriptPlugin.addResLine(line) # Add a new line to the table of the main program.
            # All others syllables
            else:
                newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")\\alpha&HFF&\\t("+sstart+","+sstart+",\\fscx200\\fscy200\\1c&HFFD700&\\alpha&H00&)"
                newsentence = newsentence + "\\t("+send+","+str(osyl[counter][4]+600)+",0.5,\\fscx100\\fscy100\\1c&HFFFFFF&)}"+syl
                line = ScriptPlugin.getAssLineOf(head, newsentence) # Reformat the line
                ScriptPlugin.addResLine(line) # Add a new line to the table of the main program.

            counter += 1
            
        # ======================================================================
        # Layer 3 : Blur
        # ======================================================================
    
        # Create a new sentence
        newsentence = ""

        # Counter
        counter = 0

        # Loop onto each syllable to create the sentence
        while counter<len(osyl):

            syl = osyl[counter][0]
            smiddle = str(osyl[counter][3]+osyl[counter][2]/2)       # middle become a string
            sstart = str(osyl[counter][3])                           # start become a string
            send = str(osyl[counter][4])                             # end become a string

            # The first syllable may contains overrides
            if counter==0:
                newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")\\fscx200\\fscy200\\blur10\\be1\\bord1\\3c&HFAE8C5&\\alpha&HFF&\\3a&H00&"
                newsentence = newsentence + "\\t("+send+","+str(osyl[counter][4]+600)+",0.5,\\fscx100\\fscy100\\3a&HFF&)}"+syl
                line = ScriptPlugin.getAssLineOf(head, newsentence) # Reformat the line
                ScriptPlugin.addResLine(line) # Add a new line to the table of the main program.
            # All others syllables
            else:
                newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")\\alpha&HFF&\\3c&HFAE8C5&\\blur10\\be1\\bord1"
                newsentence = newsentence + "\\t("+sstart+","+sstart+",\\fscx200\\fscy200\\3a&H00&)"
                newsentence = newsentence + "\\t("+send+","+str(osyl[counter][4]+600)+",0.5,\\fscx100\\fscy100\\3a&HFF&)}"+syl
                line = ScriptPlugin.getAssLineOf(head, newsentence) # Reformat the line
                ScriptPlugin.addResLine(line) # Add a new line to the table of the main program.

            counter += 1
            
        # ======================================================================
        # Layer 4 : Drawings (lightning)
        # ======================================================================
    
        # Create a new sentence
        newsentence = ""

        # Counter
        counter = 0
        
        sfx = ["a","b","c","d"]

        # Loop onto each syllable to create the sentence
        while counter<len(osyl):

            syl = osyl[counter][0]
            smiddle = str(osyl[counter][3]+osyl[counter][2]/2)       # middle become a string
            sstart = str(osyl[counter][3])                           # start become a string
            send = str(osyl[counter][4])                             # end become a string

            # The first syllable may contains overrides
            if counter==0:
                kounter = 20
                layer = 0
                loopfx = 0
                
                # str(kounter) represents start - str(kounter+20) represents end - period is 20ms
                # There is at each 20ms a new lightning until the end of the time of the syllable
                while kounter<osyl[counter][2]:
                    
                    layer += 1
                    newhead = ScriptPlugin.phChangeLayer(head, str(layer))                    
                    
                    if kounter==0:                        
                        newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")"
                        newsentence = newsentence + "\\fscx200\\fscy200\\fnPowerOfGod\\bord1\\3a&H7F&\\1a&H7F&\\3c&HFFFFFF&"
                        newsentence = newsentence + "\\t("+str(kounter+20)+","+str(kounter+20)+",\\alpha&HFF&)}"+sfx[loopfx]
                    else:
                        newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")"                        
                        newsentence = newsentence + "\\fscx200\\fscy200\\fnPowerOfGod\\bord1\\3a&H7F&\\1a&H7F&\\3c&HFFFFFF&"
                        newsentence = newsentence + "\\frz"+str(kounter)+"\\alpha&HFF&"
                        newsentence = newsentence + "\\t("+str(kounter)+","+str(kounter)+",\\3a&H7F&\\1a&H7F&)"
                        newsentence = newsentence + "\\t("+str(kounter+20)+","+str(kounter+20)+",\\alpha&HFF&)}"+sfx[loopfx]
                    kounter += 20
                    
                    line = ScriptPlugin.getAssLineOf(newhead, newsentence) # Reformat the line
                    ScriptPlugin.addResLine(line) # Add a new line to the table of the main program.
                    
                    loopfx += 1
                    if loopfx==4:
                        loopfx = 0
                    
            # All others syllables
            else:
                kounter = 20
                layer = 0
                loopfx = 0
                
                # str(kounter) represents start - str(kounter+20) represents end - period is 20ms
                # There is at each 20ms a new lightning until the end of the time of the syllable
                while kounter<osyl[counter][2]:
                    
                    layer += 1
                    newhead = ScriptPlugin.phChangeLayer(head, str(layer))                    
                    
                    if kounter==0:                        
                        newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")"
                        newsentence = newsentence + "\\fscx200\\fscy200\\fnPowerOfGod\\bord1\\3a&H7F&\\1a&H7F&\\3c&HFFFFFF&"
                        newsentence = newsentence + "\\t("+str(osyl[counter][3]+kounter+20)+","+str(osyl[counter][3]+kounter+20)+",\\alpha&HFF&)}"+sfx[loopfx]
                    else:
                        newsentence = "{\\an5\\pos("+str(pxArray[counter]+leftstartpos)+","+posy+")"                        
                        newsentence = newsentence + "\\fscx200\\fscy200\\fnPowerOfGod\\bord1\\3a&H7F&\\1a&H7F&\\3c&HFFFFFF&"
                        newsentence = newsentence + "\\frz"+str(kounter)+"\\alpha&HFF&"
                        newsentence = newsentence + "\\t("+str(osyl[counter][3]+kounter)+","+str(osyl[counter][3]+kounter)+",\\3a&H7F&\\1a&H7F&)"
                        newsentence = newsentence + "\\t("+str(osyl[counter][3]+kounter+20)+","+str(osyl[counter][3]+kounter+20)+",\\alpha&HFF&)}"+sfx[loopfx]
                    kounter += 20
                    
                    line = ScriptPlugin.getAssLineOf(newhead, newsentence) # Reformat the line
                    ScriptPlugin.addResLine(line) # Add a new line to the table of the main program.
                    
                    loopfx += 1
                    if loopfx==4:
                        loopfx = 0

            counter += 1

File: src/source_v0.2_Java6-8/SmallBoxForFansub/docs/advkaraoke.rb
# To change this template, choose Tools | Templates
# and open the template in the editor.

# This line is very important if you want to use java code in ruby.
# You must write it at every time.
require 'java'

# Register the function into the main program.
Java::smallboxforfansub.scripting.ScriptPlugin.rubyRegister("Up and down","upanddown","1.0","A simple wave","AssFxMaker owner (Chien-Rouge)","","")

def upanddown
  # Get a selected line (line = head + sentence)
  line = Java::smallboxforfansub.scripting.ScriptPlugin.getSelectedOrgLine()
  
  # Get the sentence for this line
  sentence = Java::smallboxforfansub.scripting.ScriptPlugin.getSentence(line);

  # Get a table of syllable parameters for the basic karaoke
  # [i][0] : syllable                   string
  # [i][1] : hundredth                  integer
  # [i][2] : thousandth                 integer
  # [i][3] : thousandth (start)         integer
  # [i][4] : thousandth (end)           integer
  osyl = Java::smallboxforfansub.scripting.ScriptPlugin.phKaraoke(sentence);

  # Create a new sentence
  newsentence = ""

  # Loop onto each syllable to create the sentence
  for i in 0..osyl.length-1 do

    syl = osyl[i][0]
    smiddle = (osyl[i][3]+osyl[i][2]/2).to_s()    # middle become a string
    sstart = osyl[i][3].to_s()                    # start become a string
    send = osyl[i][4].to_s()                      # end become a string

    # The first syllable may contains overrides
    if i==0
      newsentence = "{\\an8\\t("+sstart+","+smiddle+",\\fscx200\\fscy200)"+"\\t("+smiddle+","+send+",\\fscx100\\fscy100)}"+syl
    # All others syllables
    else
      newsentence = newsentence + "{\\r\\t("+sstart+","+smiddle+",\\fscx200\\fscy200)"+"\\t("+smiddle+","+send+",\\fscx100\\fscy100)}"+syl
    end
  end

  # Reformat the line
  head = Java::smallboxforfansub.scripting.ScriptPlugin.getHead(line);
  line = Java::smallboxforfansub.scripting.ScriptPlugin.getAssLineOf(head, newsentence);
  
  # Add a new line to the table of the main program.
  Java::smallboxforfansub.scripting.ScriptPlugin.addResLine(line)
end

# Register the function into the main program.
Java::smallboxforfansub.scripting.ScriptPlugin.rubyRegister("Blue to green","bluetogreen","1.0","","AssFxMaker owner (Chien-Rouge)","","")

def bluetogreen
  
  # It's better to load the 'tools.rb' script by using the absolute path.
  # Here is a method to obtain the path without you have to change something else.
  # But if you prefer then type your absolute path :
  # Example on Windows : require "C:\\Programs\\AssFxMaker\\Effects\\tools.rb"
  require Java::smallboxforfansub.scripting.ScriptPlugin.getScriptsPath()+"tools.rb"

  # Get a selected line (line = head + sentence)
  line = Java::smallboxforfansub.scripting.ScriptPlugin.getSelectedOrgLine()

  # Get the sentence for this line
  sentence = Java::smallboxforfansub.scripting.ScriptPlugin.getSentence(line);

  # Get a table of syllable parameters for the basic karaoke
  # [i][0] : syllable                   string
  # [i][1] : hundredth                  integer
  # [i][2] : thousandth                 integer
  # [i][3] : thousandth (start)         integer
  # [i][4] : thousandth (end)           integer
  osyl = Java::smallboxforfansub.scripting.ScriptPlugin.phKaraoke(sentence);

  # Get a gradation of colors (these functions are in the 'tools.rb' script)
  @colors = gradation_of_colors("HFF3333","H33CC00",count_syllables(sentence))
  @invert = gradation_of_colors(invert_color("HFF3333"),invert_color("H33CC00"),count_syllables(sentence))

  # Create a new sentence
  newsentence = ""

  # Loop onto each syllable to create the sentence
  for i in 0..osyl.length-1 do

    syl = osyl[i][0]
    smiddle = (osyl[i][3]+osyl[i][2]/2).to_s()    # middle become a string
    sstart = osyl[i][3].to_s()                    # start become a string
    sextra = (osyl[i][4]+500).to_s()
    scolor = @colors[i].to_s()
    sinvert = @invert[i].to_s()

    # The first syllable may contains overrides
    if i==0
      newsentence = "{\\an8\\1c&"+sinvert+"&\\3c&HFFFFFF&\\t("+sstart+","+smiddle+",\\bord7\\blur7)"+"\\t("+smiddle+","+sextra+",\\bord2\\blur0\\1c&"+scolor+"&)}"+syl
    # All others syllables
    else
      newsentence = newsentence + "{\\r\\1c&"+scolor+"&\\3c&HFFFFFF&\\t("+sstart+","+sstart+",\\1c&"+sinvert+"&)\\t("+sstart+","+smiddle+",\\bord7\\blur7)"+"\\t("+smiddle+","+sextra+",\\bord2\\blur0\\1c&"+scolor+"&)}"+syl
    end
  end

  # Reformat the line
  head = Java::smallboxforfansub.scripting.ScriptPlugin.getHead(line);
  line = Java::smallboxforfansub.scripting.ScriptPlugin.getAssLineOf(head, newsentence);

  # Add a new line to the table of the main program.
  Java::smallboxforfansub.scripting.ScriptPlugin.addResLine(line)
end

# Register the function into the main program.
Java::smallboxforfansub.scripting.ScriptPlugin.rubyRegister("Small grad (array mode)","smallgradation","1.0","","AssFxMaker owner (Chien-Rouge)","","")

def smallgradation
  # Load of the 'tools.rb' library.
  require Java::smallboxforfansub.scripting.ScriptPlugin.getScriptsPath()+"tools.rb"

  # Get all selected lines (line = head + sentence)
  @lines = Java::smallboxforfansub.scripting.ScriptPlugin.getSelectedOrgLines()
  
  # Use a dialog of AssFxMaker.
  dialog = Java::smallboxforfansub.scripting.AssColorDialog
  acd = dialog.new(nil,true)
  acd.setLocationRelativeTo(nil)
  
  # Set up the color 1 of the gradation
  color_one = acd.showDialog("FFFFFF")
  
  # Set up the color 2 of the gradation
  color_two = acd.showDialog("FFFFFF")
  
  # For each 'line' in @lines
  @lines.each do |line|
    # Get the sentence for this line
	sentence = Java::smallboxforfansub.scripting.ScriptPlugin.getSentence(line)
	
	# Get a table of syllable parameters for the basic karaoke
	# [i][0] : syllable                   string
	# [i][1] : hundredth                  integer
    # [i][2] : thousandth                 integer
    # [i][3] : thousandth (start)         integer
    # [i][4] : thousandth (end)           integer
    osyl = Java::smallboxforfansub.scripting.ScriptPlugin.phKaraoke(sentence)
	
	# Get a gradation of colors (these functions are in the 'tools.rb' script)
	@border_colors = gradation_of_colors("H"+color_one,"H"+color_two,count_syllables(sentence))
	@text_colors = gradation_of_colors("H"+color_two,"H"+color_one,count_syllables(sentence))
	
	# Create a new sentence
    newsentence = ""
	
	# Loop onto each syllable to create the sentence
    for i in 0..osyl.length-1 do

      syl = osyl[i][0]
      smiddle = (osyl[i][3]+osyl[i][2]/2).to_s()    # middle become a string
      sstart = osyl[i][3].to_s()                    # start become a string
      sextra = (osyl[i][4]+500).to_s()
      sbcolor = @border_colors[i].to_s()
      stcolor = @text_colors[i].to_s()

      # The first syllable may contains overrides
      if i==0
        newsentence = "{\\an8\\org(-10000,640)\\1c&"+stcolor+"&\\3c&"+sbcolor+"&\\t("+sstart+","+smiddle+",\\bord7\\blur7\\frz0.02\\1c&HFFFFFF&)"+"\\t("+smiddle+","+sextra+",\\bord2\\blur0\\frz0\\1c&"+stcolor+"&)}"+syl
      # All others syllables
      else
        newsentence = newsentence + "{\\r\\1c&"+stcolor+"&\\3c&"+sbcolor+"&\\t("+sstart+","+smiddle+",\\bord7\\blur7\\frz0.02\\1c&HFFFFFF&)"+"\\t("+smiddle+","+sextra+",\\bord2\\blur0\\frz0\\1c&"+stcolor+"&)}"+syl
      end
    end

    # Reformat the line
    head = Java::smallboxforfansub.scripting.ScriptPlugin.getHead(line);
    line = Java::smallboxforfansub.scripting.ScriptPlugin.getAssLineOf(head, newsentence);

    # Add a new line to the table of the main program.
    Java::smallboxforfansub.scripting.ScriptPlugin.addResLine(line)	
  end
end

# Register the function into the main program.
Java::smallboxforfansub.scripting.ScriptPlugin.rubyRegister("Wide grad (array mode)","widegradation","1.0","","AssFxMaker owner (Chien-Rouge)","","")

def widegradation
  # Load of the 'tools.rb' library.
  require Java::smallboxforfansub.scripting.ScriptPlugin.getScriptsPath()+"tools.rb"

  # Get all selected lines (line = head + sentence).
  @lines = Java::smallboxforfansub.scripting.ScriptPlugin.getSelectedOrgLines()
  
  # Ask for the width of video (you can define it in the option (default -> 1280)).
  video_width = Java::smallboxforfansub.scripting.ScriptPlugin.getVideoWidth()
  
  # Use a dialog of AssFxMaker.
  dialog = Java::smallboxforfansub.scripting.InputDialog
  id = dialog.new(nil,true)
  id.setLocationRelativeTo(nil)
  
  # Set up the ordinate 1 of the gradation
  id.setDialogMessage("Choose the ordinate to start the gradation.")
  id.setDialogTitle("Choose a number.")
  y_one = id.showDialog("20")
  
  # Set up the ordinate 2 of the gradation
  id.setDialogMessage("Choose the ordinate to end the gradation.")
  id.setDialogTitle("Choose a number.")
  y_two = id.showDialog("70")
  
  # Use a dialog of AssFxMaker.
  dialog = Java::smallboxforfansub.scripting.AssColorDialog
  acd = dialog.new(nil,true)
  acd.setLocationRelativeTo(nil)
  
  # Set up the color 1 of the gradation
  color_one = acd.showDialog("FFFFFF")
  
  # Set up the color 2 of the gradation
  color_two = acd.showDialog("0000FF")
  
  # For each 'line' in @lines
  @lines.each do |line|
    # Get the sentence for this line
	sentence = Java::smallboxforfansub.scripting.ScriptPlugin.getSentence(line)
	
	# Get a table of syllable parameters for the basic karaoke
	# [i][0] : syllable                   string
	# [i][1] : hundredth                  integer
    # [i][2] : thousandth                 integer
    # [i][3] : thousandth (start)         integer
    # [i][4] : thousandth (end)           integer
    osyl = Java::smallboxforfansub.scripting.ScriptPlugin.phKaraoke(sentence)
	
	# Get a gradation of colors (these functions are in the 'tools.rb' script)
	@border_colors = gradation_of_colors("H"+color_one,"H"+color_two,count_syllables(sentence))
	@pixel_by_pixel = gradation_of_colors("H"+color_one,"H"+color_two,y_two.to_i()-y_one.to_i())
	
	# Create a new sentence
    newsentence = ""
	
	# Loop from y_one to y_two
	for j in y_one.to_i()..y_two.to_i() do
	  # Loop onto each syllable to create the sentence
      for i in 0..osyl.length-1 do

        syl = osyl[i][0]
        smiddle = (osyl[i][3]+osyl[i][2]/2).to_s()    # middle become a string
        sstart = osyl[i][3].to_s()                    # start become a string
        sextra = (osyl[i][4]+500).to_s()
        sbcolor = @border_colors[i].to_s()
        stcolor = @pixel_by_pixel[j-y_one.to_i()].to_s()

        # The first syllable may contains overrides
        if i==0
          newsentence = "{\\an8\\clip(0,"+j.to_s()+","+video_width.to_s()+","+(j+1).to_s()+")\\org(-10000,640)\\1c&"+stcolor+"&\\3c&"+sbcolor+"&\\t("+sstart+","+smiddle+",\\bord7\\blur7\\frz0.02\\1c&HFFFFFF&)"+"\\t("+smiddle+","+sextra+",\\bord2\\blur0\\frz0\\1c&"+stcolor+"&)}"+syl
        # All others syllables
        else
          newsentence = newsentence + "{\\r\\1c&"+stcolor+"&\\3c&"+sbcolor+"&\\t("+sstart+","+smiddle+",\\bord7\\blur7\\frz0.02\\1c&HFFFFFF&)"+"\\t("+smiddle+","+sextra+",\\bord2\\blur0\\frz0\\1c&"+stcolor+"&)}"+syl
        end
      end

      # Get the head
      head = Java::smallboxforfansub.scripting.ScriptPlugin.getHead(line)
	  
	  # Change the layer
	  head = Java::smallboxforfansub.scripting.ScriptPlugin.phChangeLayer(head, j.to_s())
	  
	  # Reformat the line
      line = Java::smallboxforfansub.scripting.ScriptPlugin.getAssLineOf(head, newsentence)

      # Add a new line to the table of the main program.
      Java::smallboxforfansub.scripting.ScriptPlugin.addResLine(line)
	end
  end
end
